packages/backend/src/QueryEngine.ts

```diff
@@ -131,7 +148,34 @@ export class QueryEngine<DB> {
                 connectionString: config.url,
                 max: 10,
             });
-        this.middlewares = config.middlewares ?? [];
+        this.middlewares = [
+            ...(config.middlewares ?? []),
+            middleware<AnyQuery, AnyContext>({
+                predicate: ({ query, context }) => {
+                    const permissions: string[] = [];
+
+                    mapRecursive(query, (node) => {
+                        if (isQueryWithPermissions(node)) {
+                            permissions.push(...(node?.permissions ?? []));
+                        }
+
+                        return node;
+                    });
+
+                    if (
+                        config.dangerouslyAllowNoPermissions ||
```

Comment:
I would move this check upwards: just don't include the middleware at all if `dangerouslyAllowNoPermissions` is true.